File: apps/web/src/lib/sync/sync-manager.ts
import { indexedDB } from '@/lib/db/indexed-db';
import { createClient } from '@/lib/supabase/client';
import type { BodyMetrics, UserProfile } from '@/types/body-metrics';
import type { DailyMetrics } from '@/lib/db/indexed-db';

export type SyncStatus = 'idle' | 'syncing' | 'success' | 'error';

interface SyncState {
  isSyncing: boolean;
  lastSyncDate: Date | null;
  syncStatus: SyncStatus;
  pendingSyncCount: number;
  error?: string;
}

class SyncManager {
  private state: SyncState = {
    isSyncing: false,
    lastSyncDate: null,
    syncStatus: 'idle',
    pendingSyncCount: 0,
  };

  private listeners: Set<(state: SyncState) => void> = new Set();
  private syncInterval: NodeJS.Timeout | null = null;
  private isOnline = typeof navigator !== 'undefined' ? navigator.onLine : true;

  constructor() {
    // Only setup browser-specific features on client side
    if (typeof window !== 'undefined') {
      // Listen for online/offline events
      window.addEventListener('online', this.handleOnline);
      window.addEventListener('offline', this.handleOffline);
      
      // Start periodic sync
      this.startPeriodicSync();
      
      // Update pending count on initialization
      this.updatePendingCount();
    }
  }

  // State management
  private updateState(updates: Partial<SyncState>) {
    this.state = { ...this.state, ...updates };
    this.notifyListeners();
  }

  private notifyListeners() {
    this.listeners.forEach(listener => listener(this.state));
  }

  subscribe(listener: (state: SyncState) => void) {
    this.listeners.add(listener);
    // Immediately notify with current state
    listener(this.state);
    
    // Return unsubscribe function
    return () => {
      this.listeners.delete(listener);
    };
  }

  // Network handlers
  private handleOnline = () => {
    this.isOnline = true;
    this.syncIfNeeded();
  };

  private handleOffline = () => {
    this.isOnline = false;
  };

  // Periodic sync
  private startPeriodicSync() {
    // Sync every 5 minutes when online
    this.syncInterval = setInterval(() => {
      if (this.isOnline) {
        this.syncIfNeeded();
      }
    }, 5 * 60 * 1000);
  }

  async syncIfNeeded() {
    if (!this.isOnline || this.state.isSyncing) return;

    const unsynced = await indexedDB.getUnsyncedItems();
    const totalUnsynced = 
      unsynced.bodyMetrics.length + 
      unsynced.dailyMetrics.length + 
      unsynced.profiles.length;

    if (totalUnsynced > 0) {
      await this.syncAll();
    }
  }

  async syncAll() {
    if (this.state.isSyncing || !this.isOnline) return;

    this.updateState({ isSyncing: true, syncStatus: 'syncing' });

    try {
      const supabase = createClient();
      const { data: { user } } = await supabase.auth.getUser();
      
      if (!user) {
        throw new Error('User not authenticated');
      }

      const unsynced = await indexedDB.getUnsyncedItems();
      let hasErrors = false;

      // Sync profiles
      for (const profile of unsynced.profiles) {
        try {
          await this.syncProfile(profile, user.id);
        } catch (error) {
          hasErrors = true;
          console.error('Profile sync error:', error);
        }
      }

      // Sync body metrics
      for (const metrics of unsynced.bodyMetrics) {
        try {
          await this.syncBodyMetrics(metrics, user.id);
        } catch (error) {
          hasErrors = true;
          console.error('Body metrics sync error:', error);
        }
      }

      // Sync daily metrics
      for (const metrics of unsynced.dailyMetrics) {
        try {
          await this.syncDailyMetrics(metrics, user.id);
        } catch (error) {
          hasErrors = true;
          console.error('Daily metrics sync error:', error);
        }
      }

      this.updateState({
        isSyncing: false,
        lastSyncDate: new Date(),
        syncStatus: hasErrors ? 'error' : 'success',
        error: hasErrors ? 'Some items failed to sync' : undefined,
      });

    } catch (error) {
      this.updateState({
        isSyncing: false,
        syncStatus: 'error',
        error: error instanceof Error ? error.message : 'Sync failed',
      });
    }

    await this.updatePendingCount();
  }

  private async syncProfile(profile: UserProfile, userId: string) {
    const supabase = createClient();
    
    const { error } = await supabase
      .from('profiles')
      .upsert({
        ...profile,
        id: userId,
        updated_at: new Date().toISOString(),
      });

    if (error) throw error;

    await indexedDB.markAsSynced('profiles', profile.id || userId);
    await indexedDB.updateSyncStatus('profiles', profile.id || userId, 'synced');
  }

  private async syncBodyMetrics(metrics: BodyMetrics, userId: string) {
    const supabase = createClient();
    
    const { error } = await supabase
      .from('body_metrics')
      .upsert({
        ...metrics,
        user_id: userId,
        updated_at: new Date().toISOString(),
      });

    if (error) throw error;

    await indexedDB.markAsSynced('bodyMetrics', metrics.id);
    await indexedDB.updateSyncStatus('bodyMetrics', metrics.id, 'synced');
  }

  private async syncDailyMetrics(metrics: DailyMetrics, userId: string) {
    const supabase = createClient();
    
    const { error } = await supabase
      .from('daily_metrics')
      .upsert({
        ...metrics,
        user_id: userId,
        updated_at: new Date().toISOString(),
      });

    if (error) throw error;

    await indexedDB.markAsSynced('dailyMetrics', metrics.id);
    await indexedDB.updateSyncStatus('dailyMetrics', metrics.id, 'synced');
  }

  async updatePendingCount() {
    const unsynced = await indexedDB.getUnsyncedItems();
    const count = 
      unsynced.bodyMetrics.length + 
      unsynced.dailyMetrics.length + 
      unsynced.profiles.length;

    this.updateState({ pendingSyncCount: count });
  }

  // Public methods for data operations
  async logWeight(weight: number, unit: string, notes?: string) {
    const supabase = createClient();
    const { data: { user } } = await supabase.auth.getUser();
    
    if (!user) throw new Error('User not authenticated');

    const metrics: BodyMetrics = {
      id: crypto.randomUUID(),
      user_id: user.id,
      date: new Date().toISOString(),
      weight,
      weight_unit: unit as 'kg' | 'lbs',
      notes,
      created_at: new Date().toISOString(),
      updated_at: new Date().toISOString(),
    };

    // Save to IndexedDB
    await indexedDB.saveBodyMetrics(metrics, user.id);
    await this.updatePendingCount();

    // Attempt immediate sync if online
    if (this.isOnline) {
      this.syncIfNeeded();
    }

    return metrics;
  }

  async logDailyMetrics(steps?: number, notes?: string) {
    const supabase = createClient();
    const { data: { user } } = await supabase.auth.getUser();
    
    if (!user) throw new Error('User not authenticated');

    const metrics: DailyMetrics = {
      id: crypto.randomUUID(),
      user_id: user.id,
      date: new Date(),
      steps,
      notes,
      created_at: new Date(),
      updated_at: new Date(),
    };

    // Save to IndexedDB
    await indexedDB.saveDailyMetrics(metrics);
    await this.updatePendingCount();

    // Attempt immediate sync if online
    if (this.isOnline) {
      this.syncIfNeeded();
    }

    return metrics;
  }

  async getLocalBodyMetrics(from?: Date, to?: Date): Promise<BodyMetrics[]> {
    const supabase = createClient();
    const { data: { user } } = await supabase.auth.getUser();
    
    if (!user) return [];

    return indexedDB.getBodyMetrics(user.id, from, to);
  }

  async getLocalDailyMetrics(date: Date): Promise<DailyMetrics | null> {
    const supabase = createClient();
    const { data: { user } } = await supabase.auth.getUser();
    
    if (!user) return null;

    return indexedDB.getDailyMetrics(user.id, date);
  }

  // Cleanup
  destroy() {
    if (this.syncInterval) {
      clearInterval(this.syncInterval);
    }
    
    window.removeEventListener('online', this.handleOnline);
    window.removeEventListener('offline', this.handleOffline);
    
    this.listeners.clear();
  }

  // Clear all local data (for logout)
  async clearAllData() {
    await indexedDB.clearAllData();
    this.updateState({
      lastSyncDate: null,
      syncStatus: 'idle',
      pendingSyncCount: 0,
      error: undefined,
    });
  }
}

export const syncManager = new SyncManager();

File: apps/ios/LogYourBody/Services/SyncManager.swift
//
// SyncManager.swift
// LogYourBody
//
import Foundation
import Combine
import Network
import Clerk

@MainActor
class SyncManager: ObservableObject {
    static let shared = SyncManager()
    
    @Published var isSyncing = false
    @Published var lastSyncDate: Date?
    @Published var syncStatus: SyncStatus = .idle
    @Published var pendingSyncCount = 0
    
    private let coreDataManager = CoreDataManager.shared
    private let authManager = AuthManager.shared
    private let supabaseManager = SupabaseManager.shared
    private let networkMonitor = NWPathMonitor()
    private let syncQueue = DispatchQueue(label: "com.logyourbody.sync", qos: .background)
    
    private var syncTimer: Timer?
    private var cancellables = Set<AnyCancellable>()
    
    enum SyncStatus {
        case idle
        case syncing
        case success
        case error(String)
    }
    
    private init() {
        setupNetworkMonitoring()
        setupAutoSync()
        updatePendingSyncCount()
    }
    
    private func setupNetworkMonitoring() {
        let queue = DispatchQueue.global(qos: .background)
        networkMonitor.start(queue: queue)
        
        networkMonitor.pathUpdateHandler = { [weak self] path in
            if path.status == .satisfied {
                // Network is available, attempt sync
                Task { @MainActor [weak self] in
                    self?.syncIfNeeded()
                }
            }
        }
    }
    
    private func setupAutoSync() {
        // Sync every 5 minutes when app is active
        syncTimer = Timer.scheduledTimer(withTimeInterval: 300, repeats: true) { [weak self] _ in
            Task { @MainActor [weak self] in
                self?.syncIfNeeded()
            }
        }
        
        // Sync when user logs in
        authManager.$currentUser
            .sink { [weak self] user in
                if user != nil {
                    Task { @MainActor [weak self] in
                        self?.syncAll()
                    }
                }
            }
            .store(in: &cancellables)
    }
    
    func syncIfNeeded() {
        Task.detached { [weak self] in
            guard let self = self else { return }
            
            guard self.networkMonitor.currentPath.status == .satisfied else {
                // print("📵 Sync skipped: No network connection")
                return
            }
            guard await self.authManager.isAuthenticated else {
                // print("🔒 Sync skipped: Not authenticated")
                return
            }
            guard await !self.isSyncing else {
                // print("⏳ Sync skipped: Already syncing")
                return
            }
            
            // Log the current user ID
            let userId = await self.authManager.currentUser?.id
            // print("🔄 SyncManager: Checking sync for user: \(userId ?? "nil")")
            
            // Check if we synced recently (within last 5 minutes)
            let lastSyncKey = "lastSupabaseSyncDate"
            if let lastSync = UserDefaults.standard.object(forKey: lastSyncKey) as? Date {
                let minutesSinceLastSync = Date().timeIntervalSince(lastSync) / 60
                if minutesSinceLastSync < 5 {
                    // print("⏰ Sync skipped: Synced \(Int(minutesSinceLastSync)) minutes ago")
                    return
                }
            }
            
            let unsynced = self.coreDataManager.fetchUnsyncedEntries()
            let totalUnsynced = unsynced.bodyMetrics.count + unsynced.dailyMetrics.count + unsynced.profiles.count
            
            // print("📊 Unsynced items: \(unsynced.bodyMetrics.count) body metrics, \(unsynced.dailyMetrics.count) daily metrics, \(unsynced.profiles.count) profiles")
            
            if totalUnsynced > 0 {
                // print("🚀 Starting sync for \(totalUnsynced) items...")
                UserDefaults.standard.set(Date(), forKey: lastSyncKey)
                await self.syncAll()
            } else {
                // print("✅ Everything is already synced")
                await self.updatePendingSyncCount()
            }
        }
    }
    
    func syncAll() {
        guard !isSyncing else { return }
        guard authManager.isAuthenticated else { return }
        
        Task { @MainActor in
            self.isSyncing = true
            self.syncStatus = .syncing
        }
        
        syncQueue.async { [weak self] in
            Task { @MainActor [weak self] in
                guard let self = self else { return }
                
                // Get the Clerk session token
                guard let session = self.authManager.clerkSession else {
                    self.isSyncing = false
                    self.syncStatus = .error("No active session")
                    return
                }
                
                do {
                    // Get JWT token from Clerk session (using new integration pattern)
                    // print("🔑 Getting Clerk session token for Supabase...")
                    let tokenResource = try await session.getToken()
                    guard let token = tokenResource?.jwt else {
                        // print("❌ Failed to get Clerk session token")
                        self.isSyncing = false
                        self.syncStatus = .error("Failed to get authentication token")
                        return
                    }
                    
                    // print("✅ Got Clerk session token, starting sync...")
                    self.performSync(token: token)
                } catch {
                    // print("❌ Token error: \(error)")
                    self.isSyncing = false
                    self.syncStatus = .error("Token error: \(error.localizedDescription)")
                }
            }
        }
    }
    
    private func performSync(token: String) {
        Task.detached { [weak self] in
            guard let self = self else { return }
            do {
                let unsynced = coreDataManager.fetchUnsyncedEntries()
                var hasErrors = false
                
                // print("📤 Starting sync: \(unsynced.bodyMetrics.count) body metrics, \(unsynced.dailyMetrics.count) daily metrics")
                
                // Sync body metrics in batches
                if !unsynced.bodyMetrics.isEmpty {
                    // print("🔍 Processing \(unsynced.bodyMetrics.count) unsynced body metrics...")
                    let bodyMetricsBatch = unsynced.bodyMetrics.compactMap { cached -> [String: Any]? in
                        guard let userId = cached.userId,
                              let id = cached.id,
                              let date = cached.date else {
                            // print("⚠️ Skipping body metric with missing data: userId=\(cached.userId ?? "nil"), id=\(cached.id ?? "nil"), date=\(String(describing: cached.date))")
                            return nil
                        }
                        
                        // Always include ALL fields, even if null, to satisfy Supabase's "All object keys must match" requirement
                        var metrics: [String: Any] = [
                            "id": id,
                            "user_id": userId,
                            "date": ISO8601DateFormatter().string(from: date),
                            "created_at": ISO8601DateFormatter().string(from: cached.createdAt ?? Date()),
                            "updated_at": ISO8601DateFormatter().string(from: cached.updatedAt ?? Date()),
                            "weight": cached.weight > 0 ? cached.weight : NSNull(),
                            "weight_unit": cached.weightUnit ?? "kg",
                            "body_fat_percentage": cached.bodyFatPercentage > 0 ? cached.bodyFatPercentage : NSNull(),
                            "body_fat_method": cached.bodyFatMethod ?? NSNull(),
                            "muscle_mass": cached.muscleMass > 0 ? cached.muscleMass : NSNull(),
                            "bone_mass": cached.boneMass > 0 ? cached.boneMass : NSNull(),
                            "notes": cached.notes ?? NSNull(),
                            "photo_url": cached.photoUrl ?? NSNull(),
                            "data_source": cached.dataSource ?? "Manual",
                            "original_photo_url": cached.originalPhotoUrl ?? NSNull(),
                            "photo_processed_at": cached.photoProcessedAt != nil ? ISO8601DateFormatter().string(from: cached.photoProcessedAt!) : NSNull()
                        ]
                        
                        return metrics
                    }
                    
                    // print("📊 After filtering: \(bodyMetricsBatch.count) body metrics ready to sync")
                    
                    // Send in smaller batches to avoid timeouts and debug issues
                    let batchSize = 50
                    var successCount = 0
                    
                    for i in stride(from: 0, to: bodyMetricsBatch.count, by: batchSize) {
                        let endIndex = min(i + batchSize, bodyMetricsBatch.count)
                        let batch = Array(bodyMetricsBatch[i..<endIndex])
                        
                        // print("📦 Sending batch \(i / batchSize + 1) of \((bodyMetricsBatch.count + batchSize - 1) / batchSize): \(batch.count) items")
                        
                        do {
                            let result = try await supabaseManager.upsertBodyMetricsBatch(batch, token: token)
                            successCount += result.count
                            // print("✅ Batch successful: \(result.count) items")
                        } catch {
                            // print("❌ Batch failed: \(error)")
                            // Continue with next batch even if this one fails
                        }
                    }
                    
                    if successCount > 0 {
                        // print("✅ Total synced: \(successCount) body metrics")
                        
                        // Mark synced items
                        for cached in unsynced.bodyMetrics {
                            if let id = cached.id {
                                coreDataManager.markAsSynced(entityName: "CachedBodyMetrics", id: id)
                            }
                        }
                    } else if !bodyMetricsBatch.isEmpty {
                        // print("❌ Failed to sync any body metrics")
                        hasErrors = true
                    }
                }
                
                // Sync daily metrics in batches
                if !unsynced.dailyMetrics.isEmpty {
                    // print("🔍 Processing \(unsynced.dailyMetrics.count) unsynced daily metrics...")
                    let dailyMetricsBatch = unsynced.dailyMetrics.compactMap { cached -> [String: Any]? in
                        guard let userId = cached.userId,
                              let id = cached.id,
                              let date = cached.date else { return nil }
                        
                        var metrics: [String: Any] = [
                            "id": id,
                            "user_id": userId,
                            "date": ISO8601DateFormatter().string(from: date),
                            "created_at": ISO8601DateFormatter().string(from: cached.createdAt ?? Date()),
                            "updated_at": ISO8601DateFormatter().string(from: cached.updatedAt ?? Date())
                        ]
                        
                        if cached.steps > 0 {
                            metrics["steps"] = Int(cached.steps)
                        }
                        
                        if let notes = cached.notes {
                            metrics["notes"] = notes
                        }
                        
                        return metrics
                    }
                    
                    // print("📊 After filtering: \(dailyMetricsBatch.count) daily metrics ready to sync")
                    
                    // Send in smaller batches to avoid timeouts
                    let batchSize = 50 // Reduced from unlimited to 50 per batch
                    var successCount = 0
                    
                    for i in stride(from: 0, to: dailyMetricsBatch.count, by: batchSize) {
                        let endIndex = min(i + batchSize, dailyMetricsBatch.count)
                        let batch = Array(dailyMetricsBatch[i..<endIndex])
                        
                        // print("📦 Sending daily metrics batch \(i / batchSize + 1) of \((dailyMetricsBatch.count + batchSize - 1) / batchSize): \(batch.count) items")
                        
                        do {
                            let result = try await supabaseManager.upsertDailyMetricsBatch(batch, token: token)
                            successCount += result.count
                            // print("✅ Batch successful: \(result.count) items")
                            
                            // Mark synced items in this batch
                            let syncedIds = Set(result.compactMap { $0["id"] as? String })
                            for cached in unsynced.dailyMetrics {
                                if let id = cached.id, syncedIds.contains(id) {
                                    coreDataManager.markAsSynced(entityName: "CachedDailyMetrics", id: id)
                                }
                            }
                        } catch {
                            // print("❌ Daily metrics batch failed: \(error)")
                            // Continue with next batch even if this one fails
                        }
                    }
                    
                    if successCount > 0 {
                        // print("✅ Total synced: \(successCount) daily metrics")
                    } else if !dailyMetricsBatch.isEmpty {
                        // print("❌ Failed to sync any daily metrics")
                        hasErrors = true
                    }
                }
                
                await MainActor.run {
                    self.isSyncing = false
                    self.lastSyncDate = Date()
                    self.syncStatus = hasErrors ? .error("Some items failed to sync") : .success
                    self.updatePendingSyncCount()
                }
            } catch {
                // print("❌ Sync error: \(error)")
                await MainActor.run {
                    self.isSyncing = false
                    self.syncStatus = .error(error.localizedDescription)
                }
            }
        }
    }
    
    private func syncProfile(_ cached: CachedProfile, token: String) async -> Bool {
        guard let url = URL(string: "\(Constants.baseURL)/api/users/profile") else { return false }
        
        var request = URLRequest(url: url)
        request.httpMethod = "PUT"
        request.setValue("Bearer \(token)", forHTTPHeaderField: "Authorization")
        request.setValue("application/json", forHTTPHeaderField: "Content-Type")
        
        let profile = cached.toUserProfile()
        
        do {
            let data = try JSONEncoder().encode(profile)
            request.httpBody = data
            
            let (_, response) = try await URLSession.shared.data(for: request)
            
            if let httpResponse = response as? HTTPURLResponse,
               httpResponse.statusCode == 200 {
                await Task { @MainActor in
                    coreDataManager.markAsSynced(entityName: "CachedProfile", id: cached.id ?? "")
                    coreDataManager.updateSyncStatus(
                        entityName: "CachedProfile",
                        id: cached.id ?? "",
                        status: "synced"
                    )
                }.value
                return true
            } else {
                await Task { @MainActor in
                    coreDataManager.updateSyncStatus(
                        entityName: "CachedProfile",
                        id: cached.id ?? "",
                        status: "error",
                        error: "HTTP error"
                    )
                }.value
                return false
            }
        } catch {
            await Task { @MainActor in
                coreDataManager.updateSyncStatus(
                    entityName: "CachedProfile",
                    id: cached.id ?? "",
                    status: "error",
                    error: error.localizedDescription
                )
            }.value
            return false
        }
    }
    
    private func syncBodyMetrics(_ cached: CachedBodyMetrics, token: String) async -> Bool {
        guard let url = URL(string: "\(Constants.baseURL)/api/weights") else { return false }
        
        var request = URLRequest(url: url)
        request.httpMethod = "PUT" // Always use PUT for upsert behavior
        request.setValue("Bearer \(token)", forHTTPHeaderField: "Authorization")
        request.setValue("application/json", forHTTPHeaderField: "Content-Type")
        
        guard let metrics = cached.toBodyMetrics() else {
            // print("⚠️ Skipping sync for corrupted body metric")
            return false
        }
        
        do {
            let encoder = JSONEncoder()
            encoder.dateEncodingStrategy = .iso8601
            let data = try encoder.encode(metrics)
            request.httpBody = data
            
            let (_, response) = try await URLSession.shared.data(for: request)
            
            if let httpResponse = response as? HTTPURLResponse {
                if (200...299).contains(httpResponse.statusCode) || httpResponse.statusCode == 409 {
                    await Task { @MainActor in
                        coreDataManager.markAsSynced(entityName: "CachedBodyMetrics", id: cached.id ?? "")
                        coreDataManager.updateSyncStatus(
                            entityName: "CachedBodyMetrics",
                            id: cached.id ?? "",
                            status: "synced"
                        )
                    }.value
                    return true
                } else {
                    await Task { @MainActor in
                        coreDataManager.updateSyncStatus(
                            entityName: "CachedBodyMetrics",
                            id: cached.id ?? "",
                            status: "error",
                            error: "HTTP \(httpResponse.statusCode)"
                        )
                    }.value
                    return false
                }
            } else {
                return false
            }
        } catch {
            await Task { @MainActor in
                coreDataManager.updateSyncStatus(
                    entityName: "CachedBodyMetrics",
                    id: cached.id ?? "",
                    status: "error",
                    error: error.localizedDescription
                )
            }.value
            return false
        }
    }
    
    private func syncDailyMetrics(_ cached: CachedDailyMetrics, token: String) async -> Bool {
        guard let url = URL(string: "\(Constants.baseURL)/api/daily-metrics") else { return false }
        
        var request = URLRequest(url: url)
        request.httpMethod = cached.createdAt == cached.updatedAt ? "POST" : "PUT"
        request.setValue("Bearer \(token)", forHTTPHeaderField: "Authorization")
        request.setValue("application/json", forHTTPHeaderField: "Content-Type")
        
        let metrics = cached.toDailyMetrics()
        
        do {
            let data = try JSONEncoder().encode(metrics)
            request.httpBody = data
            
            let (_, response) = try await URLSession.shared.data(for: request)
            
            if let httpResponse = response as? HTTPURLResponse,
               (200...299).contains(httpResponse.statusCode) {
                await Task { @MainActor in
                    coreDataManager.markAsSynced(entityName: "CachedDailyMetrics", id: cached.id ?? "")
                    coreDataManager.updateSyncStatus(
                        entityName: "CachedDailyMetrics",
                        id: cached.id ?? "",
                        status: "synced"
                    )
                }.value
                return true
            } else {
                await Task { @MainActor in
                    coreDataManager.updateSyncStatus(
                        entityName: "CachedDailyMetrics",
                        id: cached.id ?? "",
                        status: "error",
                        error: "HTTP error"
                    )
                }.value
                return false
            }
        } catch {
            await Task { @MainActor in
                coreDataManager.updateSyncStatus(
                    entityName: "CachedDailyMetrics",
                    id: cached.id ?? "",
                    status: "error",
                    error: error.localizedDescription
                )
            }.value
            return false
        }
    }
    
    func updatePendingSyncCount() {
        Task.detached { [weak self] in
            guard let self = self else { return }
            let unsynced = self.coreDataManager.fetchUnsyncedEntries()
            let count = unsynced.bodyMetrics.count + unsynced.dailyMetrics.count + unsynced.profiles.count
            
            await MainActor.run {
                self.pendingSyncCount = count
            }
        }
    }
    
    // MARK: - Public Methods for Manual Operations
    
    func logWeight(_ weight: Double, unit: String, notes: String? = nil) {
        guard let userId = authManager.currentUser?.id else { return }
        
        let id = UUID().uuidString
        let now = Date()
        
        let metrics = BodyMetrics(
            id: id,
            userId: userId,
            date: now,
            weight: weight,
            weightUnit: unit,
            bodyFatPercentage: nil,
            bodyFatMethod: nil,
            muscleMass: nil,
            boneMass: nil,
            notes: notes,
            photoUrl: nil,
            dataSource: "Manual",
            createdAt: now,
            updatedAt: now
        )
        
        // Save to Core Data
        coreDataManager.saveBodyMetrics(metrics, userId: userId)
        
        // Update pending count
        updatePendingSyncCount()
        
        // Attempt immediate sync if online
        syncIfNeeded()
    }
    
    // Check if weight entry exists for a specific date
    func weightEntryExists(for date: Date) -> Bool {
        guard let userId = authManager.currentUser?.id else { return false }
        
        let calendar = Calendar.current
        // Check within a 1-hour window to handle minor time differences
        let hourBefore = calendar.date(byAdding: .hour, value: -1, to: date) ?? date
        let hourAfter = calendar.date(byAdding: .hour, value: 1, to: date) ?? date
        
        return !coreDataManager.fetchBodyMetrics(
            for: userId,
            from: hourBefore,
            to: hourAfter
        ).isEmpty
    }
    
    // Save weight entry from HealthKit
    func saveWeightEntry(_ entry: WeightEntry) async throws {
        guard let userId = authManager.currentUser?.id else { return }
        
        let id = UUID().uuidString
        let now = Date()
        
        let metrics = BodyMetrics(
            id: id,
            userId: userId,
            date: entry.loggedAt,
            weight: entry.weight,
            weightUnit: entry.weightUnit,
            bodyFatPercentage: nil,
            bodyFatMethod: nil,
            muscleMass: nil,
            boneMass: nil,
            notes: entry.notes,
            photoUrl: nil,
            dataSource: "Manual",
            createdAt: now,
            updatedAt: now
        )
        
        // Save to Core Data
        coreDataManager.saveBodyMetrics(metrics, userId: userId)
        
        // Update pending count
        updatePendingSyncCount()
        
        // Attempt immediate sync if online
        syncIfNeeded()
    }
    
    // Save complete body metrics (including body fat)
    func logBodyMetrics(_ metrics: BodyMetrics) {
        guard let userId = authManager.currentUser?.id else { return }
        
        // Ensure the metrics have the correct user ID
        let metricsWithUserId = BodyMetrics(
            id: metrics.id,
            userId: userId,
            date: metrics.date,
            weight: metrics.weight,
            weightUnit: metrics.weightUnit,
            bodyFatPercentage: metrics.bodyFatPercentage,
            bodyFatMethod: metrics.bodyFatMethod,
            muscleMass: metrics.muscleMass,
            boneMass: metrics.boneMass,
            notes: metrics.notes,
            photoUrl: metrics.photoUrl,
            dataSource: metrics.dataSource,
            createdAt: metrics.createdAt,
            updatedAt: metrics.updatedAt
        )
        
        // Save to Core Data
        coreDataManager.saveBodyMetrics(metricsWithUserId, userId: userId)
        
        // Update pending count
        updatePendingSyncCount()
        
        // Attempt immediate sync if online
        syncIfNeeded()
    }
    
    // Check if daily metrics exist for a specific date
    func dailyMetricsExists(for date: Date) -> Bool {
        guard let userId = authManager.currentUser?.id else { return false }
        
        return coreDataManager.fetchDailyMetrics(for: userId, date: date) != nil
    }
    
    // Save daily metrics from HealthKit
    func saveDailyMetrics(steps: Int, date: Date, notes: String? = nil) async throws {
        guard let userId = authManager.currentUser?.id else { return }
        
        let id = UUID().uuidString
        let now = Date()
        
        let metrics = DailyMetrics(
            id: id,
            userId: userId,
            date: date,
            steps: steps,
            notes: notes,
            createdAt: now,
            updatedAt: now
        )
        
        // Save to Core Data
        coreDataManager.saveDailyMetrics(metrics, userId: userId)
        
        // Update pending count
        updatePendingSyncCount()
        
        // Attempt immediate sync if online
        syncIfNeeded()
    }
    
    func logDailyMetrics(steps: Int?, notes: String? = nil) {
        guard let userId = authManager.currentUser?.id else { return }
        
        let id = UUID().uuidString
        let now = Date()
        
        let metrics = DailyMetrics(
            id: id,
            userId: userId,
            date: now,
            steps: steps,
            notes: notes,
            createdAt: now,
            updatedAt: now
        )
        
        // Save to Core Data
        coreDataManager.saveDailyMetrics(metrics, userId: userId)
        
        // Update pending count
        updatePendingSyncCount()
        
        // Attempt immediate sync if online
        syncIfNeeded()
    }
    
    func fetchLocalBodyMetrics(from startDate: Date? = nil, to endDate: Date? = nil) -> [BodyMetrics] {
        guard let userId = authManager.currentUser?.id else { return [] }
        
        let cached = coreDataManager.fetchBodyMetrics(for: userId, from: startDate, to: endDate)
        return cached.compactMap { $0.toBodyMetrics() }
    }
    
    func fetchLocalDailyMetrics(for date: Date) -> DailyMetrics? {
        guard let userId = authManager.currentUser?.id else { return nil }
        
        return coreDataManager.fetchDailyMetrics(for: userId, date: date)?.toDailyMetrics()
    }
}
